core/domain/classifier_services.py

```diff
@@ -100,8 +101,39 @@ def classify_string_classifier_rule(state, normalized_answer):
             training_examples.extend([
                 [doc, [str(answer_group_index)]]
                 for doc in classifier_rule_spec.inputs['training_data']])
-    if len(training_examples) > 0:
+
+    exploration = exp_services.get_exploration_by_id(exp_id)
+    state = exploration.states[state_name]
+    algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+        'TextInput']['classifier_id']
+    if state.classifier_model_id:
+        classifier = get_classifier_by_id(state.classifier_model_id)
+        sc_dict = classifier.cached_classifier_data
+        sc = classifier_registry.Registry.get_classifier_by_algorithm_id(
+            algorithm_id)
+        sc.from_dict(sc_dict)
```

Comment:
Yes, that's right. One thing to consider is taking a look at the other domain objects -- something seems fishy about this one, and it may be a long-standing issue that's worth fixing now before it spreads. In particular, if you look at the from_dict() methods in exp_domain.py, you'll notice that they generally call the constructor and create a new class with the desired properties.

So, my question is, what should be the lifecycle of a Classifier object? Should we:

- initialize an empty object and then modify it to add params, or
- initialize an object by giving it the necessary params at the outset?

In thinking about how to answer this question, I go back to the user flows we want to support. I think there are two (but it's been a while since I looked at this in depth, so please correct me if I'm wrong):

- Loading a classifier from a pre-saved model
- Training a classifier from scratch, based on training data.

With the former, we would ideally just provide the necessary params at the outset and have from_dict call the constructor with those params. With the latter, I wonder if it makes sense to have a from_training_data method that does the training immediately and initializes the classifier with the output parameters, rather than having this slightly awkward initial stage where the classifier is "empty". Is there a good reason, based on existing code calls, to separate classifier initialization from train()?

Finally, I just wanted to say that we might not need to do all of this in this PR. But I think it's good for both of us to have a clear idea of what the desired "end state" is, so that we can make changes that gradually move towards that.